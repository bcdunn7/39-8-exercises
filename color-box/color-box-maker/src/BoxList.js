import { v4 as uuid } from 'uuid';
import { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

function BoxList() {
    const remove = (id) => {
        setBoxes(boxes.filter(b => b.id !== id))
    };
    const INITIAL_STATE = []
    const [boxes, setBoxes] = useState(INITIAL_STATE);

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {...newBox, id: uuid()}])
    }

    return (
        <>
            <NewBoxForm addBox={addBox}/>
            {boxes.map(b => (
                <Box 
                    key={b.id}
                    id={b.id}
                    color={b.color} 
                    height={b.height} 
                    width={b.width} 
                    remove={remove}
                />
            ))}
        </>
    );
}
  
export default BoxList;
