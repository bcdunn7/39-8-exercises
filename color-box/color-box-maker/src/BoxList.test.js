import React from "react";
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import BoxList from "./BoxList";

test("it renders without crashing", function() {
	render(<BoxList/>);
});

test("it matches snapshot", () => {
    const { asFragment } = render(<BoxList/>);
    expect(asFragment()).toMatchSnapshot();
})

test("can add new item", () => {
    const { getByLabelText, queryByText } = render(<BoxList/>);

    // no box (and therefore no button with 'X')
    expect(queryByText("X")).not.toBeInTheDocument();

    const colorInput = getByLabelText("Background Color:");
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const submitBtn = queryByText("Add!");

    // form
    fireEvent.change(colorInput, { target: {value: "red" }})
    fireEvent.change(widthInput, { target: {value: "30" }})
    fireEvent.change(heightInput, { target: {value: "30" }})
    fireEvent.click(submitBtn)

    // box exists
    expect(queryByText("X")).toBeInTheDocument();
})

test("can remove item", () => {
    const { getByLabelText, queryByText } = render(<BoxList/>);

    const colorInput = getByLabelText("Background Color:");
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const submitBtn = queryByText("Add!");

    fireEvent.change(colorInput, { target: {value: "red" }})
    fireEvent.change(widthInput, { target: {value: "30" }})
    fireEvent.change(heightInput, { target: {value: "30" }})
    fireEvent.click(submitBtn)

    // box exists
    expect(queryByText("X")).toBeInTheDocument();

    // click remove button
    const removeBtn = queryByText("X");
    fireEvent.click(removeBtn)

    // box does not exist
    expect(queryByText("X")).not.toBeInTheDocument();
})
