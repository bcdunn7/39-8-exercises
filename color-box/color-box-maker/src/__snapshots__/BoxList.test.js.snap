// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Background Color:
    </label>
    <input
      id="color"
      name="color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="text"
      value=""
    />
    <button>
      Add!
    </button>
  </form>
</DocumentFragment>
`;
